cmd/clusterctl/pkg/client/cluster/inventory.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog"
+	clusterctlv1 "sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/config"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/internal/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// InventoryClient exposes methods to interface with a cluster's provider inventory.
+type InventoryClient interface {
+	// EnsureCRD installs the CRD required for creating inventory items, if necessary.
+	// Nb. In order to provide a simpler out-of-the box experience, the inventory CRD
+	// is embedded in the clusterctl binary.
+	EnsureCRD() error
+
+	// Validate checks the inventory in order to determine if adding a new provider instance to the cluster.
+	// can lead to an non functional cluster.
+	Validate(clusterctlv1.Provider) error
+
+	// Create an inventory item for a provider instance installed in the cluster.
+	Create(clusterctlv1.Provider) error
+
+	// Return inventory items for all the provider instances installed in the cluster.
+	List() ([]clusterctlv1.Provider, error)
+}
+
+// inventoryClient implements InventoryClient.
+type inventoryClient struct {
+	proxy Proxy
+}
+
+// ensure inventoryClient implements InventoryClient.
+var _ InventoryClient = &inventoryClient{}
+
+// newInventoryClient returns a inventoryClient.
+func newInventoryClient(proxy Proxy) *inventoryClient {
+	return &inventoryClient{
+		proxy: proxy,
+	}
+}
+
+func (p *inventoryClient) EnsureCRD() error {
+	// get the CRD manifest from the embedded assets
+	yaml, err := config.Asset("cmd/clusterctl/config/manifest/clusterctl-api.yaml")
```

Comment:
nit: consider putting the path as a const at the top of the file.

Reply:
done